### ISSUE TITLE
Release resource for Python and Java worker

### ISSUE THREAD

**Issue by surehb** [ray-project/ray#2454]
Currently, both Python and Java worker write a bunch of records into Redis but we don't have a mechanism to release them afterwards; in PR #2406 , we introduced a dictionary that maps from driver id to serialization context, but we won't delete the item after the driver exits or dies. 

In order to solve above issues, we propose some ideas here: [design doc](https://docs.google.com/document/d/1UbCT23GDN1rZeCXY97HG6dM3jQ3xGU-Y0p-V8zuWrwE/edit?usp=sharing), please help to comment. Thanks in advance!

**Comment by surehb** [comment]
@robertnishihara, @eric-jj, FYI.

**Comment by eric-jj** [comment]
@robertnishihara @pcmoritz  do you have comments on the design? If you don't have concern, we will go on execution base on the design.

**Comment by robertnishihara** [comment]
I'm looking at it now.

**Comment by robertnishihara** [comment]
Thanks @surehb @eric-jj, this is a great idea and so far we have been neglecting it! I have a few high-level comments.

1. I don't think driver heartbeats are necessary, since we can detect whenever a driver disconnects from the node manager (and we can also store the mapping from driver ID to node manager ID).
2. Any new information that is added to the GCS should be done through the "tables" interface (unless there is a good reason not to). This provides a unified way to store/retrieve information from Redis and will be shared between Python/Java. See https://github.com/ray-project/ray/blob/master/src/ray/gcs/tables.h. Stuff that is currently in pure Python can stay that way for now (but should eventually be migrated).
3. About the implementation. The code for "adding" stuff to the GCS and the code for "removing" stuff from the GCS are coupled in the sense that they each have to have knowledge of the underlying schema. Therefore, it makes sense for both of these functionalities to be implemented in the same place. The "table" API is the natural way to do this, but for the stuff that is in pure Python, adding to and removing from a specific table should possibly be part of the same class.
4. Some of the schemas will probably need to be redesigned in the future.
5. Lastly, let's implement all of this functionality in multiple PRs. In particular, I think the following can be done separately
    - Detecting driver deaths. We started working on this in https://github.com/ray-project/ray/pull/2289, but it looks like it didn't get finished.
    - Doing cleanups for the exports (also related to #2289).
    - Doing cleanups for the task table and object table (also related to #2289).
    - Other cleanups.

cc @stephanie-wang @elibol

**Comment by eric-jj** [comment]
@robertnishihara Thank you for your comments.
We planned to use the heartbeat because we haven't noticed there are existing one in the node manager. But there is another question as implement it in the node manager. In our previous design, the scan and cleanup logical is in the GCS module, so there is an instance to do that. In the node manager to do that means we need select 1 node manager to do that, do you have an idea how to select it?

**Comment by robertnishihara** [comment]
I'd suggest doing the cleanup in the "monitor" process.

Currently we have two monitors (I assume down the road we will just have one). @stephanie-wang does GCS cleanup after a driver exits belong in the Python monitor or the C++ monitor?

Currently that functionality is all in the Python monitor, see https://github.com/ray-project/ray/blob/7edc677304f90bcab0a01e9722e59e36dd9afcb0/python/ray/monitor.py#L429-L436 (though note that that function is entirely for legacy Ray).

@elibol was working on an equivalent for x-ray, see #2289, and I think some of that approach is necessary to address this issue.

Basically, the pattern is
1. The driver dies/exits.
2. The node manager broadcasts a message that the driver died.
3. The monitor receives the message and cleans up the GCS entries for that driver.

**Comment by eric-jj** [comment]
@robertnishihara I thought about put the logical to monitor before. But it is tricky to put business logical in a monitor service. Which means the monitor become a service on critical path. We did the similar design in our system before, but we have to add a new service monitor_monitor to detect the monitor works well or not.

**Comment by robertnishihara** [comment]
Hm that's an interesting point.

If we do GCS clean up in the node manager, the natural node manager to do the cleanup would be the one that detected the driver death. This of course doesn't handle the case where the driver died because the node failed, but that may be less important.

However, GCS cleanup can be very expensive (e.g., iterating through the task table and object table), so if it's done in the node manager it would potentially need to be a separate thread.

**Comment by surehb** [comment]
@robertnishihara, thanks very much for your suggestions!

I agree that we should use "tables" interface. That was the design actually, since the API does a bound of things together on Redis (add, publish, clean up). Sorry that I forgot to describe these details in the spec.

The idea of heartbeat is based on these thoughts:
    -- The machine may die, as you mentioned;
    -- The driver may hang or exit ungracefully, in that case node manager won't know the death.
The side impact of heartbeat is it will increase Redis load, but as node manager already use this way to find each other (heartbeat interval is 1 second), and we can actually adjust the interval used for drivers, I would think it is acceptable in general.
Cleaning up on node manager cannot work for machine failure case, which should be a valid concern in real production usage. Another side impact is doing in monitor process, we will have to create multiple redis connections during the process, but if we do in redis module, the overall performance should be better, right?

I have replied your comments in the spec, you can also take a look.

**Comment by surehb** [comment]
@robertnishihara, I have relied in the spec about your question. I know that currently in _clean_up_entries_for_driver we scan the Redis to find out the records that belong to the specific driver which is super heavy. In order to address this issue, we introduced a separate set in Redis which will store all the keys that belongs to a driver, then can do clean up efficiently.

**Comment by elibol** [comment]
@surehb Having a set of all the keys that belong to a driver sounds good to me. We ought to update the current implementation, and the implementation in #2289 once these changes have been merged.

**Comment by robertnishihara** [comment]
I'm a little skeptical of having a set of keys that belong to a driver. For example, if this is done as part of the driver table, then this would double the number of Redis commands that we have to do, right (every time you add a key, you also have to update the set)?

Alternatively, if each Redis shard keeps its own set for each driver and we rely on custom Redis modules, then it will be much harder to replace Redis with other stores in the future (if we want to).

As for driver heartbeats, there are a lot of ways this could go wrong. E.g., if the user's application code does something for 5 seconds without releasing the GIL then the driver could be accidentally marked dead. But primarily, we already rely on node manager to detect driver deaths when they disconnect and to clean up that way, and I'd prefer to have a single mechanism to do this instead of two.

**Comment by surehb** [comment]
Thanks @robertnishihara and @elibol for raising all the valid points, we are getting closer to a good solution!

We are now having two questions under discussion:
1. Driver death detection.
The main problem in previous solution is it cannot handle driver hang and machine failure, I believe we have to solve it in production level system. To deal with driver hand case, we cannot rely on the driver itself to notify its death; to deal with machine failure, we cannot rely on the local node manager. Heartbeat to store (Redis for now) seems to be the only choice as far as I can see. GIL is indeed a valid concern. I guess we can solve it in this way: adjust the detection policy, for example set heartbeat period to 5 seconds, and consider a driver dead if lose heartbeat in 60 seconds. Given Python has a fixed interval for a thread to release GIL, see [here](https://realpython.com/python-gil/). 
2. Redis clean up.
In order to improve the perf we introduce a set that stores all the keys of resources belong to a driver. Yes, this will double the Redis command number. But in the previous solution, we scan the all the tables, get each item out and check if it belongs to a specific driver, I believe it is much more heavier. Note that, all will be integrated in custom Redis module so only one command is called from the client's perspective. 
As for potential store replacement, this is a good point to consider about. But since all existing interfaces are in custom module, we will have re-write this part if we want to move to new stores, and clean up is just a small part of this effort, right?
Theoretically, it make sense to wrapper the clean up actions for a driver into one command inside custom Redis module, since it saves redis connections compared with doing separately from a client. The question here is where do we trigger this command, or a client (monitor process) or inside Redis.

I haven't considered about sharding carefully, I will find a time to think about it.

**Comment by eric-jj** [comment]
@robertnishihara @raulchen The GIL is not a problem for the heartbeat. The heartbeat thread will be in c++ module, which can shared by the python and java worker. 
And for the duplicate keys in redis, it is a problem. How about to let the python code to subscribe the set belongs to the driver, which can make the logical more reasonable.

**Comment by surehb** [comment]
Hi @robertnishihara, 
For death detection, does is sound good to you if we put heartbeat into GCS client (c++)?
For Redis clean up, if you still have concern about storing all the keys separately, then maybe we should store all the things that belong to a driver under one key? That will need to redesign the schema in Redis.

**Comment by robertnishihara** [comment]
@surehb, let's start with the approach of detecting driver deaths by detecting when the client disconnects from the node manager (this is in progress in #2289, although @elibol is pretty busy right now so I'm not sure what the timeline on that is). We can wait and see if the driver hanging but not disconnecting problem is an issue.

As for Redis clean up, we are working pretty hard to not use any special Redis features and to just use a very generic log interface. So I'd like to prioritize cleanup without worrying about efficiency, and then consider efficiency optimizations later.

**Comment by robertnishihara** [comment]
@surehb @eric-jj I have a question about "hanging" drivers. How much does this happen in practice? It seems like it shouldn't be very common. Also, even if a driver is hanging (i.e., it is an infinite loop or blocked somewhere), it would still continue to send heartbeats, so driver heartbeats wouldn't help detect its failure, right?

**Comment by surehb** [comment]
@robertnishihara, I don't have an exact number at hand. Driver hang may happen anywhere including heartbeat module (probably GCS client). If a driver lose the connection with outside, we should treat it as dead. If it continues to send heartbeat, then we will do clean up when driver timeout (heartbeat stops). Actually driver hang is only one of the cases, if we consider all possible cases including driver hang, driver ungraceful exit (I still don't understand why this case can be handled, please see my reply in the shared spec above.) and machine level failure, that could be a pain point based on my experience about using a similar platform.

**Comment by raulchen** [comment]
Another thing that needs to be cleaned up is `sys.path`, see https://github.com/ray-project/ray/issues/2327#issuecomment-409836904.

**Comment by ericl** [comment]
Auto-closing stale issue.
